File: binding_rv_adapter/src/main/java/me/lx/rv/loadmore22/LoadMoreAdapter.java
//package me.lx.rv.loadmore22;
//
//import android.database.Observable;
//import android.util.Log;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//
//import androidx.annotation.NonNull;
//import androidx.recyclerview.widget.GridLayoutManager;
//import androidx.recyclerview.widget.LinearLayoutManager;
//import androidx.recyclerview.widget.RecyclerView;
//import androidx.recyclerview.widget.StaggeredGridLayoutManager;
//
//import java.lang.reflect.Field;
//import java.util.ArrayList;
//import java.util.List;
//
//import me.lx.rv.loadmore.LoadMoreFooter;
//import me.lx.rv.loadmore.LoadMoreViewHolder;
//
//@SuppressWarnings({"unchecked", "WeakerAccess", "unused"})
//public class LoadMoreAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
//
//    /**
//     *
//     */
//    private RecyclerView.Adapter<RecyclerView.ViewHolder> mAdapter;
//    /**
//     *
//     */
//    private int VIEW_TYPE_LOAD_MORE = 1111;
//    /**
//     *
//     */
//    private OnLoadMoreListener mOnLoadMoreListener;
//    /**
//     *
//     */
//    private OnFailedClickListener mOnFailedClickListener;
//    /**
//     * 是否为上拉
//     */
//    private boolean mIsScrollLoadMore = false;
//
//    /**
//     * footer的状态
//     */
//    static final int STATE_LOADING = 0;
////    static final int STATE_LOAD_COMPLETE = 1;
//    static final int STATE_LOAD_FAILED = 2;
//    static final int STATE_NO_MORE_DATA = 3;
//    private int mStateType = STATE_LOADING;
//
//    /**
//     * 已无更多数据
//     */
//    private boolean mNoMoreData = false;
//
//    /**
//     *
//     */
//    private RecyclerView mRecyclerView;
//    /**
//     *
//     */
//    private AbsLoadMoreFooter mFooter;
//
//    public static LoadMoreAdapter wrap(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter) {
//        return new LoadMoreAdapter(adapter);
//    }
//
//    public static LoadMoreAdapter wrap(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter, AbsLoadMoreFooter footer) {
//        return new LoadMoreAdapter(adapter, footer);
//    }
//
//    private LoadMoreAdapter(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter) {
//        this(adapter, new LoadMoreFooter());
//    }
//
//    private LoadMoreAdapter(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter, AbstractMethodError footer) {
//        if (adapter == null) {
//            throw new NullPointerException("mAdapter can not be null");
//        }
//        this.mAdapter = adapter;
//        this.mFooter = footer;
//    }
//
//    @Override
//    public int getItemCount() {
//        int count = mAdapter.getItemCount();
//        return count > 0 ? count + 1 : 0;
//    }
//
//    @Override
//    public long getItemId(int position) {
//        if (position == 0) return mAdapter.getItemId(position);
//        if (getItemViewType(position) == VIEW_TYPE_LOAD_MORE) return VIEW_TYPE_LOAD_MORE;
//        return mAdapter.getItemId(position);
//    }
//
//    @Override
//    public int getItemViewType(int position) {
//        if (position <= 0) return super.getItemViewType(position);
//        if (position == getItemCount() - 1) {
//            return VIEW_TYPE_LOAD_MORE;
//        }
//        return super.getItemViewType(position);
//    }
//
//    @NonNull
//    @Override
//    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
//        if (viewType == VIEW_TYPE_LOAD_MORE) {
//            final View footView = LayoutInflater.from(parent.getContext())
//                    .inflate(mFooter.setLayoutRes(), parent, false);
//            mFooter.onCreate(footView);
//            return new LoadMoreViewHolder(footView, mFooter);
//        }
//        return mAdapter.onCreateViewHolder(parent, viewType);
//    }
//
//    @Override
//    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
//
//    }
//
//    @Override
//    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, int position, @NonNull List<Object> payloads) {
//        if (holder instanceof LoadMoreViewHolder) {
//            final LoadMoreViewHolder loadMoreVH = (LoadMoreViewHolder) holder;
//            //首次如果itemView没有填充满RecyclerView，继续加载更多
//            if (!mRecyclerView.canScrollVertically(-1) && mOnLoadMoreListener != null) {
//                //fix bug Cannot call this method while RecyclerView is computing a layout or scrolling
//                mRecyclerView.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        mStateType = STATE_LOADING;
//                        loadMoreVH.setState(mStateType);
//                        mOnLoadMoreListener.onLoadMore(LoadMoreAdapter.this);
//                    }
//                });
//            }
//
//            //加载失败点击事件
//            holder.itemView.setOnClickListener(new View.OnClickListener() {
//                @Override
//                public void onClick(View v) {
//                    if (mStateType == STATE_LOAD_FAILED && mOnFailedClickListener != null) {
//                        mStateType = STATE_LOADING;
//                        mOnFailedClickListener.onClick(LoadMoreAdapter.this, holder.itemView);
//                    }
//                }
//            });
//
//            //更新状态
//            loadMoreVH.setState(mStateType);
//            Log("stateType == " + mStateType);
//        } else {
//            mAdapter.onBindViewHolder(holder, position, payloads);
//        }
//    }
//
//    @Override
//    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
//        mRecyclerView = recyclerView;
//        setFullSpan(recyclerView);
//
//        if (!isRegistered()) {
//            mAdapter.registerAdapterDataObserver(dataObserver);
//        }
//
//        recyclerView.addOnScrollListener(mOnScrollListener);
//
//        mAdapter.onAttachedToRecyclerView(recyclerView);
//    }
//
//    @Override
//    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
//        mRecyclerView = null;
//
//        if (isRegistered()) {
//            mAdapter.unregisterAdapterDataObserver(dataObserver);
//        }
//
//        recyclerView.removeOnScrollListener(mOnScrollListener);
//
//        mAdapter.onDetachedFromRecyclerView(recyclerView);
//    }
//
//    @Override
//    public void onViewAttachedToWindow(@NonNull RecyclerView.ViewHolder holder) {
//        if (holder instanceof LoadMoreViewHolder) return;
//        mAdapter.onViewAttachedToWindow(holder);
//    }
//
//    @Override
//    public void onViewDetachedFromWindow(@NonNull RecyclerView.ViewHolder holder) {
//        if (holder instanceof LoadMoreViewHolder) return;
//        mAdapter.onViewDetachedFromWindow(holder);
//    }
//
//    @Override
//    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
//        if (holder instanceof LoadMoreViewHolder) return;
//        mAdapter.onViewRecycled(holder);
//    }
//
//    @Override
//    public boolean onFailedToRecycleView(@NonNull RecyclerView.ViewHolder holder) {
//        if (holder instanceof LoadMoreViewHolder) return false;
//        return mAdapter.onFailedToRecycleView(holder);
//    }
//
//    private RecyclerView.AdapterDataObserver dataObserver = new RecyclerView.AdapterDataObserver() {
//        @Override
//        public void onChanged() {
//            LoadMoreAdapter.this.notifyDataSetChanged();
//        }
//
//        @Override
//        public void onItemRangeChanged(int positionStart, int itemCount) {
//            LoadMoreAdapter.this.notifyItemRangeChanged(positionStart, itemCount);
//        }
//
//        @Override
//        public void onItemRangeChanged(int positionStart, int itemCount, @Nullable Object payload) {
//            LoadMoreAdapter.this.notifyItemRangeChanged(positionStart, itemCount, payload);
//        }
//
//        @Override
//        public void onItemRangeInserted(int positionStart, int itemCount) {
//            LoadMoreAdapter.this.notifyItemRangeInserted(positionStart, itemCount);
//        }
//
//        @Override
//        public void onItemRangeRemoved(int positionStart, int itemCount) {
//            LoadMoreAdapter.this.notifyItemRangeRemoved(positionStart, itemCount);
//        }
//
//        @Override
//        public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
//            LoadMoreAdapter.this.notifyItemRangeChanged(fromPosition, toPosition, itemCount);
//        }
//    };
//
//    private boolean isRegistered() {
//        boolean isRegistered = false;
//        try {
//            Class<? extends RecyclerView.Adapter> clazz = RecyclerView.Adapter.class;
//            Field field = clazz.getDeclaredField("mObservable");
//            field.setAccessible(true);
//            Observable observable = (Observable) field.get(mAdapter);
//
//            Field observersField = Observable.class.getDeclaredField("mObservers");
//            observersField.setAccessible(true);
//            ArrayList<Object> list = (ArrayList<Object>) observersField.get(observable);
//            isRegistered = list.contains(dataObserver);
//        } catch (Exception e) {
//            e.printStackTrace();
//        }
//        return isRegistered;
//    }
//
//    public interface OnLoadMoreListener {
//        void onLoadMore(LoadMoreAdapter adapter);
//    }
//
//    public interface OnFailedClickListener {
//        void clickLoadMoreFailView(LoadMoreAdapter adapter, View view);
//    }
//
//    public LoadMoreAdapter setLoadMoreListener(OnLoadMoreListener listener) {
//        this.mOnLoadMoreListener = listener;
//        return this;
//    }
//
//    public LoadMoreAdapter setOnFailedClickListener(OnFailedClickListener listener) {
//        this.mOnFailedClickListener = listener;
//        return this;
//    }
//
//    private RecyclerView.OnScrollListener mOnScrollListener = new RecyclerView.OnScrollListener() {
//        @Override
//        public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
//
//            if (newState != RecyclerView.SCROLL_STATE_IDLE ||
//                    mOnLoadMoreListener == null
//                    || mNoMoreData
//                    || !mIsScrollLoadMore) return;
//
//            if (canLoadMore(recyclerView.getLayoutManager())) {
//                setState(STATE_LOADING);
//                mOnLoadMoreListener.onLoadMore(LoadMoreAdapter.this);
//            }
//        }
//
//        @Override
//        public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
//            mIsScrollLoadMore = dy > 0;
//        }
//    };
//
//    private boolean canLoadMore(final RecyclerView.LayoutManager layoutManager) {
//        boolean canLoadMore = false;
//        if (layoutManager instanceof GridLayoutManager) {
//            canLoadMore = ((GridLayoutManager) layoutManager).findLastVisibleItemPosition() >= layoutManager.getItemCount() - 1;
//        } else if (layoutManager instanceof LinearLayoutManager) {
//            canLoadMore = ((LinearLayoutManager) layoutManager).findLastVisibleItemPosition() >= layoutManager.getItemCount() - 1;
//        } else if (layoutManager instanceof StaggeredGridLayoutManager) {
//            StaggeredGridLayoutManager sgLayoutManager = ((StaggeredGridLayoutManager) layoutManager);
//            int[] into = new int[sgLayoutManager.getSpanCount()];
//            int[] lastVisibleItemPositions = sgLayoutManager.findLastVisibleItemPositions(into);
//            int lastPosition = lastVisibleItemPositions[0];
//            for (int value : into) {
//                if (value > lastPosition) {
//                    lastPosition = value;
//                }
//            }
//            canLoadMore = lastPosition >= layoutManager.getItemCount() - 1;
//        }
//        return canLoadMore;
//    }
//
//    public RecyclerView.Adapter<RecyclerView.ViewHolder> getRealAdapter() {
//        return mAdapter;
//    }
//
//
////    public void loadFailed() {
////        setState(STATE_LOAD_FAILED);
////    }
////
////    public void noMoreData() {
////        mNoMoreData = true;
////        setState(STATE_NO_MORE_DATA);
////    }
//
//    private void setState(int state) {
//        if (mStateType == state) return;
//        this.mStateType = state;
//        notifyLoadMoreVH();
//    }
//
//    public void resetNoMoreData() {
//        mNoMoreData = false;
//    }
//
//    private void notifyLoadMoreVH() {
//        if (getItemCount() <= 0) return;
//        LoadMoreAdapter.this.notifyItemChanged(getItemCount() - 1);
//    }
//
//    private void setFullSpan(RecyclerView recyclerView) {
//        final RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
//        if (layoutManager == null) return;
//
//        if (layoutManager instanceof GridLayoutManager) {
//            final GridLayoutManager gm = (GridLayoutManager) layoutManager;
//            gm.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
//                @Override
//                public int getSpanSize(int position) {
//                    int viewType = getItemViewType(position);
//                    if (viewType == VIEW_TYPE_LOAD_MORE) return gm.getSpanCount();
//                    return 1;
//                }
//            });
//        } else if (layoutManager instanceof StaggeredGridLayoutManager) {
//            StaggeredGridLayoutManager sgm = (StaggeredGridLayoutManager) layoutManager;
//        }
//    }
//
//    private void Log(String text) {
//        Log.d("LoadMoreAdapter", text);
//    }
//
//
//}

File: app/src/main/java/me/lx/rv/loadmore/LoadMoreAdapter.kt
package me.lx.rv.loadmore

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import androidx.databinding.*
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import me.lx.rv.BuildConfig
import me.lx.rv.tools.Ls
import me.lx.rv.tools.WrapperUtils


/**
 *  author: luoXiong
 *  e-mail: 382060748@qq.com
 *  date: 2019/9/6 09:01
 *  version: 1.0
 *  desc:
 *  注意地方
 *  1、RecyclerView 不应该被当前类持有
 *  2、mLoadMoreListener 是在model里的,会被一直持有,所以里面的每个ob对象都得重新设置监听,设置过的得移除
 *  1、当Item里的数据使用Ob不会更新时,是因为要重写 registerAdapterDataObserver 方法
 *
 *  当有10条数据,输入1 搜索到2条,此时再输入2,没有数据的时候, 就报下面这个错误 todo 2020-1-1
 *    java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid view holder adapter positionLoadMoreViewHolder{9f83fac position=0 id=-1, oldPos=4, pLpos:4 scrap [attachedScrap] tmpDetached no parent} com.pda.work.base.view.AppRecyclerView{bfad9bc VFED..... ......I.
 *    0,0-1440,1321 #7f080138 app:id/recyclerView}, adapter:me.lx.rv.loadmore.LoadMoreAdapter@b8e2d45
 */
@Suppress("UNCHECKED_CAST")
open class LoadMoreAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private var mInflater: LayoutInflater? = null
    @LayoutRes
    protected var mLoadMoreRedId = 0


    private var mIsLoading: Boolean = false

    private var mLoadMoreListener: LoadMoreListener? = null
    private var mLoadMoreFailClickListener: LoadMoreFailClickListener? = null

    // 布局 ->单一的
    // val simpleItemBinding = itemBindingOf<Int>(me.lx.rv.R.layout.rv_load_more_layout)

    private val mDataObserver = object : RecyclerView.AdapterDataObserver() {
        override fun onChanged() {
            Ls.d("加载更多..onChanged().....11111111....")
            mIsLoading = false
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            Ls.d("加载更多..onItemRangeRemoved()..44444..noCanScroll=${noCanScrollVertically()} positionStart=$positionStart itemCount=$itemCount getCount=${mInnerAdapter.itemCount})")
            mIsLoading = false
            dataChangeNotifyLoadMoreVH()
        }

        override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
            Ls.d("加载更多..onItemRangeMoved()..55555..fromPosition=$fromPosition toPosition=$toPosition itemCount=$itemCount getCount=${mInnerAdapter.itemCount}")
            mIsLoading = false
            dataChangeNotifyLoadMoreVH()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            Ls.d("加载更多  onItemRangeInserted().......666666....")
            mIsLoading = false
            dataChangeNotifyLoadMoreVH()
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
            Ls.d("加载更多  onItemRangeChanged().......777777....")
            mIsLoading = false
        }

        // 会执行这
        override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
            mIsLoading = false
            Ls.d("加载更多 onItemRangeChanged().......3333....positionStart=${positionStart}..itemCount=${itemCount}")
        }
    }

    private lateinit var mInnerAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>
    private lateinit var mFooter: AbstractLoadMoreFooter

    private var mLastState = STATE_LOADING
    private lateinit var mRecyclerView: RecyclerView
    private var mLoadMoreHolder: LoadMoreViewHolder? = null
    private var mNoMoreData = false // 没有更多数据的标识
    private var mIfNoMoreDataHideLayout = false // 如果没有更多数据,是否隐藏底部布局
    /**
     * 是否为上拉
     */
    private var mIsScrollLoadMore = false
    private val mOnScrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
            Ls.d("onScrollStateChanged.....newState=$newState...mNoMoreData=$mNoMoreData")
            if (newState != RecyclerView.SCROLL_STATE_IDLE ||
                    mLoadMoreListener == null
                    || mNoMoreData
            //  || !mIsScrollLoadMore  // 不能用这个变量判断,当loadmore显示的时候,稍微往上移动下dy变成复数,这里就会被拦截
            ) {
                return
            }
            if (canLoadMore(recyclerView.layoutManager)) {
                //  Ls.d("监听到底部了.......loadMore()....2222222........mLastState=${mLastState}")
                setState(STATE_LOADING, 1)
                requestLoadingMore(1)
            } else {
                setState(STATE_NO_MORE_DATA, 2)
            }
        }

        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            mIsScrollLoadMore = dy > 0
        }
    }

    constructor(
            rawAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>
            //,footer: AbstractLoadMoreFooter? = null
    ) : super() {
        onConstructor(rawAdapter)
    }

    private fun onConstructor(rawAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>) {
        mInnerAdapter = rawAdapter
        mFooter = getLoadMoreFooter()
        mLoadMoreRedId = mFooter.getLayoutRes()
        // registerAdapterDataObserver(mDataObserver)
    }


    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        Ls.d("loadMore...onAttachedToRecyclerView()..111...rv=${recyclerView.hashCode()} this=${hashCode()}")
        mRecyclerView = recyclerView
        WrapperUtils.onAttachedToRecyclerView(mInnerAdapter, recyclerView, object : WrapperUtils.SpanSizeCallback {
            override fun getWrapSpanSize(
                    gridLayoutManager: GridLayoutManager,
                    oldLookup: GridLayoutManager.SpanSizeLookup?,
                    position: Int
            ): Int {
                if (position < 0) return gridLayoutManager.spanCount
                val viewType = getItemViewType(position)
                if (viewType == VIEW_TYPE_LOAD_MORE) {
                    return gridLayoutManager.spanCount
                } else if (oldLookup != null) {
                    return oldLookup.getSpanSize(position)
                } else {
                    return 1
                }
            }
        })
        recyclerView.addOnScrollListener(mOnScrollListener)
    }


    override fun getItemCount(): Int {
        val count = mInnerAdapter.itemCount
        return if (count > 0) count + 1 else 0
        // return mInnerAdapter.itemCount + if (hasLoadMore()) 1 else 0
    }

    override fun getItemViewType(position: Int): Int {
        if (position == itemCount - 1 && position != 0) {
            Ls.d("getItemViewType()....isLoadMore  itemCount=${(itemCount)}")
            return VIEW_TYPE_LOAD_MORE;
        }
        return mInnerAdapter.getItemViewType(position);
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == VIEW_TYPE_LOAD_MORE) {
            Ls.d("onCreateViewHolder()....创建LoadMoreHolder前....loadmoreIsNull=${mLoadMoreHolder == null}")
            if (mLoadMoreHolder != null) return mLoadMoreHolder!!

            if (mInflater == null) {
                mInflater = LayoutInflater.from(parent.context)
            }
            val loadMoreBinding = DataBindingUtil.inflate(
                    mInflater!!,
                    mLoadMoreRedId,
                    parent,
                    false
            ) as ViewDataBinding
            //加载失败点击事件
            loadMoreBinding.root.setOnClickListener(View.OnClickListener {
                if (mLastState == STATE_LOAD_FAILED) {
                    setState(STATE_LOADING, 3)
//                    if (mLoadMoreFailClickListener != null) {
//                        mLoadMoreFailClickListener!!.clickLoadMoreFailView(
//                            this@LoadMoreAdapter,
//                            loadMoreBinding.root
//                        )
//                    } else {
//                    }
                    requestLoadingMore(2)
                }
            })
            Ls.d("创建加载更多Holder....onCreateViewHolder()...666....isNull=${mLoadMoreHolder == null} mLastState=${getStateText(mLastState)}")
            if (mLoadMoreHolder == null) {
                mLoadMoreHolder = LoadMoreViewHolder(loadMoreBinding.root, mFooter)
                // 这里必须得更新下,当第一次加载到3条就没有更多数据了,得更新文案
                // true也是必须的,第一次创建一定要设置下数据
                setState(mLastState, 9, true) // todo 这里第3个参数是true,不需要设置吧,默认就是正常状态
            }

            return mLoadMoreHolder!!
        }
        return mInnerAdapter.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        mInnerAdapter.onBindViewHolder(holder, position)
    }

    /*
    * 1、当只有2条数据,没有铺满屏幕,执行onBindViewHolder()方法的话,
    *   先判断屏幕能否滑动,不能的话,请求加载更多,.. 只有2条数据,一般是不能滑动的
    * 2、用index记录,第1次=1,才去判断是否滑动
    * */
    private var index = 0

    @CallSuper
    override fun onBindViewHolder(
            holder: RecyclerView.ViewHolder,
            position: Int,
            payloads: List<Any>
    ) {
        if (holder is LoadMoreViewHolder) {
            if (index == 0) {
                index = 1
                if (!mRecyclerView.canScrollVertically(-1)) {
                    // 如果不能滑动,强制请求加载更多
                    requestLoadingMore(3)
                }
            }
            //首次如果itemView没有填充满RecyclerView，继续加载更多
            Ls.d("onBindViewHolder()....loadMore()...能滑动=${mRecyclerView.canScrollVertically(-1)}")
//            if (!mRecyclerView.canScrollVertically(-1) && mLoadMoreListener != null && canLoadMore(mRecyclerView.layoutManager)) {
//                //fix bug Cannot call this method while RecyclerView is computing a layout or scrolling
//                mRecyclerView.post {
//                    setState(STATE_LOADING, 4)
//                    requestLoadingMore(3)
//                }
//            }
        } else {
            mInnerAdapter.onBindViewHolder(holder, position, payloads)

        }
    }

    override fun onViewAttachedToWindow(holder: RecyclerView.ViewHolder) {
        mInnerAdapter.onViewAttachedToWindow(holder)
        if (isShowLoadMore(holder.layoutPosition)) {
            WrapperUtils.setFullSpan(holder);
        }
    }


    override fun unregisterAdapterDataObserver(observer: RecyclerView.AdapterDataObserver) {
        super.unregisterAdapterDataObserver(observer)
        //当使用ViewPager绑定View的时候会报错: IllegalStateException("Observer " + observer + " was not registered
        //  mInnerAdapter.unregisterAdapterDataObserver(observer)
    }

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        mInnerAdapter.onViewDetachedFromWindow(holder)
    }

    override fun getItemId(position: Int): Long {
        return mInnerAdapter.getItemId(position)
    }

    //
    override fun setHasStableIds(hasStableIds: Boolean) {
        mInnerAdapter.setHasStableIds(hasStableIds)
    }

    override fun onFailedToRecycleView(holder: RecyclerView.ViewHolder): Boolean {
        return mInnerAdapter.onFailedToRecycleView(holder)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        mInnerAdapter.onDetachedFromRecyclerView(recyclerView)
    }

    //
    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        mInnerAdapter.onViewRecycled(holder)
    }

    /**
     * 一定要重写该方法
     */
    override fun registerAdapterDataObserver(observer: RecyclerView.AdapterDataObserver) {
        super.registerAdapterDataObserver(observer)
        super.registerAdapterDataObserver(mDataObserver)
        // mInnerAdapter.registerAdapterDataObserver(observer)
//        Ls.d("registerAdapterDataObserver()...2222.....code=${hashCode()}")
    }

    private fun isShowLoadMore(position: Int): Boolean {
        return position >= mInnerAdapter.itemCount && mInnerAdapter.itemCount > 0
    }

    private fun dataChangeNotifyLoadMoreVH() {
        if (itemCount == 0) {
            notifyDataSetChanged()
        }
//        if (noCanScrollVertically()) {
//            notifyLoadMoreVH()
//        }
    }

    private fun noCanScrollVertically(): Boolean {
        return !mRecyclerView.canScrollVertically(-1)
    }

    private fun requestLoadingMore(flag: Int) {
        val enabledOfLoadMore = isEnabledOfLoadMore()
        Ls.d("requestLoadingMore()...触发了66....flag=$flag  itemCount=$itemCount enabled=$enabledOfLoadMore mNoMoreData=$mNoMoreData")
        if (enabledOfLoadMore) {
            mLoadMoreListener!!.isShowLoadMoreFailOb()?.set(false)
            mLoadMoreListener!!.onLoadingMore()
        }
    }

    private fun getStateText(state: Int): String {
        return when (state) {
            STATE_LOADING -> {
                "加载中"
            }
            STATE_NO_MORE_DATA -> {
                "无更多"
            }
            STATE_LOAD_FAILED -> {
                "失败"
            }
            STATE_HIDE_LAYOUT -> {
                "隐藏"
            }
            else -> {
                "未知"
            }
        }
    }

    private fun setState(state: Int, flag: Int, isForceSet: Boolean = false) {

        if (DEBUG) {
            val isShowViewOb = mLoadMoreListener!!.isShowLoadMoreViewLayout()
            val isShowView = isShowViewOb == null || isShowViewOb.get()
            Ls.d(
                    "设置加载更多状态...setState()..222...newState=${getStateText(state)} " +
                            " mLastState=${getStateText(mLastState)}  " +
                            "flag=$flag isForceSet=$isForceSet  showView=${isShowView} mNoMoreData=$mNoMoreData"
            )
        }
        if (mLastState == state && !isForceSet) return

        //if (isShowView) {
        mLastState = state
        if (mLoadMoreHolder != null) {
            // 延迟更新UI的原因是, 当加载后,添加item,刷新列表时,会看到无更多数据的脚布局,体验不好
            mLoadMoreHolder!!.setState(state)
//            mRecyclerView.postDelayed({
//                notifyLoadMoreVH()
//            }, 500)
        }
        // }
    }

    /**
     * 不清除回调会导致内存泄漏
     * 比如: LoadMoreListener是在model里的,当ViewPager有多个frag,其中第1个frag销毁了,再重新创建的时候,
     * ob还是会有之前的引用
     */
    private fun getBaseObservableClass(ob: BaseObservable): Class<BaseObservable> {
        val superclass = ob.javaClass.superclass?.superclass
        if (superclass == BaseObservable::class.java) {
            return superclass
        }
        return superclass?.superclass as Class<BaseObservable>
    }

    private fun clearObCallback(ob: BaseObservable?) {
        ob ?: return
//        Ls.d("clearObCallback()..ob=${ob.javaClass}  s1=${ob.javaClass.superclass} s2=${ob.javaClass.superclass?.superclass}")
        // 暴力反射获取属性
        val field = getBaseObservableClass(ob).getDeclaredField("mCallbacks")
        // 设置反射时取消Java的访问检查，暴力访问
        field.isAccessible = true
        val callbackObj = field.get(ob)
        if (callbackObj is PropertyChangeRegistry) {
            callbackObj.clear()
        }
    }

    fun listenerNoMoreDataAndFail(listener: LoadMoreListener): LoadMoreAdapter {
        this.mLoadMoreListener = listener
        // 防止内存泄漏.因为listener属于model,是长生命周期,当前对象是短生命的,走这个方法,得把之前对象设置的引用监听得全部移除掉
        clearObCallback(listener.isShowLoadMoreFailOb())
        clearObCallback(listener.isShowNoMoreDataOb())
        clearObCallback(listener.isShowLoadMoreViewLayout())
        val noMoreDataOb = listener.isShowNoMoreDataOb()
        if (noMoreDataOb != null) {
            mNoMoreData = noMoreDataOb.get() // 这里要赋值
            noMoreDataOb.addOnPropertyChangedCallback(object : Observable.OnPropertyChangedCallback() {
                override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                    mNoMoreData = (sender as ObservableBoolean).get()
                    Ls.d("加载更多..没有更多数据..666..onPropertyChanged..value=${mNoMoreData}")
                    if (mNoMoreData) {
                        setState(STATE_NO_MORE_DATA, 5)
                    } else {
                        setState(STATE_LOADING, 6)
                    }
                }
            })
        }
        if (mNoMoreData) {
            mLastState = STATE_NO_MORE_DATA
        }
        listener.isShowLoadMoreFailOb()
                ?.addOnPropertyChangedCallback(object : Observable.OnPropertyChangedCallback() {
                    override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                        val isShowFail = (sender as ObservableBoolean).get()
                        if (isShowFail) {
                            setState(STATE_LOAD_FAILED, 7)
                        } else {
                            setState(STATE_LOADING, 8)
                        }
                    }
                })
        val showLoadMore = listener.isShowLoadMoreViewLayout()
        if (showLoadMore != null) {
            showLoadMore
                    .addOnPropertyChangedCallback(object : Observable.OnPropertyChangedCallback() {
                        override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                            val isShowView = (sender as ObservableBoolean).get()
                            if (isShowView) {
                                setState(mLastState, 10)
                            } else {
                                setState(STATE_HIDE_LAYOUT, 11)
                            }
                        }
                    })
            showLoadMore.notifyChange()
        }

        return this
    }


    private fun notifyLoadMoreVH() {
        if (isEnabledOfLoadMore()) {
            Ls.d("noti@fyLoadMoreVH()...66666....")
            notifyItemChanged(itemCount - 1)
        }
    }

    private fun isEnabledOfLoadMore(): Boolean {
        return mLoadMoreHolder != null && itemCount > 0 && !mNoMoreData
    }

    private fun setFullSpan(recyclerView: RecyclerView) {

    }

    private fun canLoadMore(layoutManager: RecyclerView.LayoutManager?): Boolean {
        if (mLoadMoreListener?.isShowNoMoreDataOb()?.get() == true) {
            return false
        }
        var canLoadMore = false
        if (layoutManager is GridLayoutManager) {
            canLoadMore = layoutManager.findLastVisibleItemPosition() >= layoutManager.itemCount - 1
        } else if (layoutManager is LinearLayoutManager) {
            canLoadMore = layoutManager.findLastVisibleItemPosition() >= layoutManager.itemCount - 1
        } else if (layoutManager is StaggeredGridLayoutManager) {
            val into = IntArray(layoutManager.spanCount)
            val lastVisibleItemPositions = layoutManager.findLastVisibleItemPositions(into)
            var lastPosition = lastVisibleItemPositions[0]
            for (value in into) {
                if (value > lastPosition) {
                    lastPosition = value
                }
            }
            canLoadMore = lastPosition > layoutManager.itemCount
        }
        return canLoadMore
    }

    /**
     * 如果没有更多的数据后,隐藏底部没有更多数据的提示
     */
    fun setIfNoMoreDataHideLayout() {
        mIfNoMoreDataHideLayout = true
    }

    interface LoadMoreFailClickListener {
        /**
         * 正在加载更多中, 这里需求请求网络
         */
        fun clickLoadMoreFailView(loadMoreAdapter: LoadMoreAdapter, root: View)
    }

    interface LoadMoreListener {
        /**
         * 正在加载更多中, 这里需求请求网络
         */
        fun onLoadingMore()

        /**
         * 显示没有更多数据的标识
         */
        fun isShowNoMoreDataOb(): ObservableBoolean? {
            return null
        }

        /**
         * 显示加载更多失败的标识
         */
        fun isShowLoadMoreFailOb(): ObservableBoolean? {
            return null
        }

        // 是否显示加载更多的布局
        fun isShowLoadMoreViewLayout(): ObservableBoolean? {
            return null
        }
    }

    companion object {
        const val VIEW_TYPE_LOAD_MORE = 0 // 加载更多类型

        private val DEBUG = BuildConfig.DEBUG

        @JvmField
        var DEFAULT_FOOTER_PATH: String? = null
        /**
         * footer的状态
         */
        const val STATE_LOADING = 1
        const val STATE_NO_MORE_DATA = 2
        const val STATE_LOAD_FAILED = 3
        const val STATE_HIDE_LAYOUT = 4

        fun with(adapter: RecyclerView.Adapter<RecyclerView.ViewHolder>): LoadMoreAdapter {
            val loadMoreAdapter = LoadMoreAdapter(adapter)
            return loadMoreAdapter
        }

        fun getLoadMoreFooter(): AbstractLoadMoreFooter {
            var footer: AbstractLoadMoreFooter? = null
            if (!TextUtils.isEmpty(DEFAULT_FOOTER_PATH)) {
                val footerClass = Class.forName(DEFAULT_FOOTER_PATH!!)
                footer = footerClass.newInstance() as AbstractLoadMoreFooter
            }
            if (footer == null) {
                footer = LoadMoreFooter()
            }
            return footer
        }
    }
}